### PR TITLE
Schema benchmarks

### PR DESCRIPTION
**Pull request by gregsdennis**
@idstam pulling into master here.

### CODE REVIEW

File: Manatee.Json.Tests.Benchmark/TestData/schema.json [hunk 1]
```diff
@@ -0,0 +1,431 @@
...
+          }
+
+        },
+        {
+          "if": { "properties": { "@type": { "const": "ViewAccountCondition" } } },
+          "then": {
+            "prod_base": {
+              "type": [ "string", "null" ],
+              "pattern": "^[0-9\\+]+$"
+            }
```
Comment (gregsdennis) [inline Manatee.Json.Tests.Benchmark/TestData/schema.json]: @idstam you have several places in your schema that call out what look like properties but outside of a property keyword.  Is this intentional?  These would be skipped by a validator, leaving only `{}` which will validate everything as passing.

File: Manatee.Json.Tests.Benchmark/TestData/schema.json [hunk 2]
```diff
@@ -0,0 +1,431 @@
...
+      },
+      "allOf": [
+        {
+          "if": { "properties": { "@type": { "const": "SignerGroup" } } },
+          "then": {
+            "required": [ "req_sign" ],
+            "req_sign": {
+              "type": "string",
+              "pattern": "^[0-9]+$"
+            }
```
Comment (gregsdennis) [inline Manatee.Json.Tests.Benchmark/TestData/schema.json]: Here again

File: Manatee.Json.Tests.Benchmark/TestData/schema.json [hunk 3]
```diff
@@ -0,0 +1,431 @@
...
+          }
+        },
+        {
+          "if": { "properties": { "@type": { "const": "CancelGroup" } } },
+          "then": {
+            "required": [ "req_sign" ],
+            "req_sign": {
+              "type": "string",
+              "pattern": "^[0-9]+$"
+            }
```
Comment (gregsdennis) [inline Manatee.Json.Tests.Benchmark/TestData/schema.json]: Here again.

**Comment by gregsdennis** [comment]
@idstam a couple of other things about the schema (not really pertinent to this test, though) in regard to this bit here:

```json
{
  "if": { "properties": { "@type": { "const": "KycReview" } } },
  "then": {
    "required": [ "profile" ],
    "properties": {
      "prod_base": {
        "type": [ "string", "null" ],
        "pattern": "^[0-9\\+]+$"
      },
      "profile": {
        "type": "string",
        "enum": [ "Administrator", "FullAuthorization", "SignAgreements" ]
      }
    }
  }
}
```

(You have a lot of these.)

First, I don't think that you're getting what you want from the `if` keyword here.  For example, the following JSON instances will all pass that subschema:

```
{}
[]
false
null
4.5
```

Note that the object doesn't have a `@type` property, and the others aren't even objects.

The result is that since the `if` passes, the `then` will be applied.

In order to fix this, you'd need to specify `type` and `required` inside the `if`.

The second concern I have about this is that you can do this more simply with less nesting by simply merging the two subschemas:

```json
{
  "required": [ "@type", "profile" ],
  "properties": {
    "@type": { "const": "KycReview" },
    "prod_base": {
      "type": [ "string", "null" ],
      "pattern": "^[0-9\\+]+$"
    },
    "profile": {
      "type": "string",
      "enum": [ "Administrator", "FullAuthorization", "SignAgreements" ]
    }
  }
}
```

This will require `@type` to be `KycReview` *and* all of the other stuff, which ultimately is what you want:  all of the other stuff *whenever* `@type` is `KycReview`.

If you make all of these modifications, you can then wrap them in an `anyOf`.  (Since your cases are mutually exclusive by design based on the value of `@type`, `anyOf` is the choice here over `oneOf` because it's more performant.)  As soon as a passing case is found, it'll return.

**Comment by idstam** [comment]
Ah, thanks for looking into that. I'll go over the schema at work tomorrow.

/johan/

Den sön 16 feb. 2020 kl 05:07 skrev Greg Dennis <notifications@github.com>:

**Comment by gregsdennis** [comment]
@idstam have you had a chance to review my comments?

**Comment by idstam** [comment]
If you mean the ones regarding our schema here in this thread; Yes.
I kept the IF's because that gives me nice debugging info when the sender produces bad json, but otherwise your comment were spot on and very valuable.

I gave you a shout out when I published a tool the other day: https://idstam.com/blog/2020-02-22-json-schema-lab/schema.

**Comment by gregsdennis** [comment]
Thanks for the contribution, @idstam.

**Pull request merged**